### ISSUE TITLE
Roslyn should always warn about passing nullable reference to a non-nullable reference

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#36769]
**Version Used**: 

```
Branch master (21 Jun 2019)
Latest commit 898bed by Heejae Chang:
added NFW to get some data on incremental parsing bug where source si� (#36620)

* added NFW to get some data on incremental parsing bug where source size and tree size is different

* more comments
```

**Steps to Reproduce**:

```cs
#nullable enable
public class C
{
  string? field;
  public void M()
  {
    field = "";      
    M(ref field);
  }
  
  void M(ref string s1)
  {
    field = null;
    s1.ToString();
  }
}

class Program
{
  static void Main()
  {
    new C().M();
  }
}
```

**Expected Behavior**:
Warning for `M(ref field);`

**Actual Behavior**:
No warnings. The program crashes at runtime with a `NullReferenceException`.

**Notes**
Roslyn does not issue a warning in the code above because `field` does not contain a null value at the moment of an invocation. However, since it's passed by reference any changes to its value will be observable in the callee. It is always unsafe to pass a reference which might contain null values to an argument which does not accept nulls.

Note that the same crash without nullability warnings is achievable even without creating an alias since Roslyn verifies nullability at the moment of evaluating an argument instead of an invocation. E.g. the following issue has the same root cause: https://github.com/dotnet/roslyn/issues/36768

**Comment by jcouv** [comment]
This is by design. 

We used to have the stricter behavior (declared type of field must match the ref parameter exactly), but we changed it in https://github.com/dotnet/roslyn/pull/35538 to fix https://github.com/dotnet/roslyn/issues/34874

There is a trade-off here. When it comes to fields, we try to track them. It is always possible that some other code or thread modifies it, but we don't assume the worst for all those cases.
The question is which design is more useful and less annoying.

Tagging @gafter as FYI.

**Comment by TessenR** [comment]
@jcouv  Could you please clarify whether the following type inference behavior is intended then? This behavior seems related the the issue.

According to Roslyn's assumption that no other code will modify the value of a ref variable it's impossible for generic invocations to observe null values. However, Roslyn seems to use declared nullability of ref arguments for type inference. It uses inferred nullability for other non-ref arguments. Is this intended?

```cs
#nullable enable
class C
{
  void M1(string? s)
  {
    if (s == null) return;
    
    GetT(ref s).ToString(); // CS8602: Dereference of a possibly null reference.
    s.ToString(); // CS8602: Dereference of a possibly null reference.
  }
    
  void M2(string? s)
  {
    if (s == null) return;
    
    GetTNonRef(s).ToString(); // no warnings
    s.ToString(); // no warnings
  }
    
  T GetTNonRef<T>(T t) => t;
  T GetT<T>(ref T t) => t;
}
```

**Comment by Youssef1313** [comment]
@jcouv Should this be closed if it's by-design?
